livre/core/pipeline/FutureMap.h

```diff
@@ -0,0 +1,297 @@
+/* Copyright (c) 2011-2016, EPFL/Blue Brain Project
+ *                     Ahmet Bilgili <ahmet.bilgili@epfl.ch>
+ *
+ * This file is part of Livre <https://github.com/BlueBrain/Livre>
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License version 3.0 as published
+ * by the Free Software Foundation.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#ifndef _FutureMap_h_
+#define _FutureMap_h_
+
+#include <livre/core/types.h>
+#include <livre/core/pipeline/Future.h>
+
+namespace livre
+{
+
+/**
+ * The Futures classes are a wrappers to access the group of future
+ * and execute thread-safe operations on them. With a given portName (name)
+ * many futures can be associated.
+ */
+
+/**
+ * FutureMap class provides basic functionality to query state of multiple futures.
+ */
+class FutureMap
+{
+public:
+
+    /**
+     * @param futures the list of futures.
+     */
+    explicit FutureMap( const Futures& futures );
+    ~FutureMap();
+
+    /**
+     * @return the futures associated with port name.
+     */
+    Futures getFutures() const;
+
+    /**
+     * Queries if all ports are ready
+     * @return true if all port inputs are ready.
+     */
+    bool isReady() const;
+
+    /**
+     * Waits all futures
+     */
+    void wait() const;
+
+    /**
+     * Waits for any futures
+     * @return true if any future is waited.
+     */
+    bool waitForAny() const;
+
+private:
+
+    struct Impl;
+    std::shared_ptr<Impl> _impl;
+};
+
+/**
+ * PortFutures for managing ports with unique names.
```

Comment:
The main reason was, while it was evolving ( since I really started it to be a simple task paralel work graph ) I called the connections as ports between filters. On my way, I saw some similar behaviours on things and made classes out of them. But, yes it will be good to add my reasoning. I am open to suggestions for renaming.